tools/federation/src/main/java/org/eclipse/rdf4j/federated/evaluation/join/ControlledWorkerBoundJoin.java

```diff
@@ -73,27 +82,17 @@ protected void handleBindings() throws Exception {
 		TaskCreator taskCreator = null;
 		Phaser currentPhaser = phaser;
 
-		// first item is always sent in a non-bound way
-		if (!isClosed() && leftIter.hasNext()) {
-			BindingSet b = leftIter.next();
-			totalBindings++;
-			if (expr instanceof StatementTupleExpr) {
-				StatementTupleExpr stmt = (StatementTupleExpr) expr;
-				if (stmt.hasFreeVarsFor(b)) {
-					taskCreator = new BoundJoinTaskCreator(strategy, stmt);
-				} else {
-					expr = new CheckStatementPattern(stmt, queryInfo);
-					taskCreator = new CheckJoinTaskCreator(strategy, (CheckStatementPattern) expr);
-				}
-			} else if (expr instanceof FedXService) {
-				taskCreator = new FedXServiceJoinTaskCreator(strategy, (FedXService) expr);
-			} else {
-				throw new RuntimeException("Expr is of unexpected type: " + expr.getClass().getCanonicalName()
-						+ ". Please report this problem.");
+		if (submitFirstResultImmediately) {
+			// first item is always sent in a non-bound way
```

Comment:
@aschwarte10 could you confirm if this comment is still correct because it jars with how I understand the the explanation. Otherwise it looks good to me.

Reply:
when inside this if branch (i.e. inside the `submitFirstResultImmediately`)  the comment is correct. It is essentially also the same code as before - only the creation of the task creator is extracted into a helper method for reuse.

So yes: inside the if branch (which in this implementation is deactivated, but supported to be enabled in sub-classes) the first item is always sent in a non bound way.

Reply:
@JervenBolleman does this answer help? Do you want to update your "changes requested" from the review?